larq_compute_engine/core/indirect_bgemm/kernel.h

```diff
@@ -1,123 +1,195 @@
+    // Assign the zero buffer that will be used for padding.
+    zero_buffer.assign(filter_size * input_depth, 0);
+
+    for (std::int32_t output_tile_start = 0;
+         output_tile_start < tiled_output_size;
+         output_tile_start += block_size_pixels) {
+      for (std::int32_t output_tile_offset = 0;
+           output_tile_offset < block_size_pixels; output_tile_offset++) {
+        const std::int32_t output_index =
+            std::min(output_tile_start + output_tile_offset, output_size - 1);
+        const std::int32_t batch_index =
+            output_index / (output_height * output_width);
+        const std::int32_t output_x = output_index % output_width;
+        const std::int32_t output_y =
+            (output_index % (output_height * output_width)) / output_width;
+        for (std::int32_t f_y = 0; f_y < filter_height; f_y++) {
+          const std::int32_t input_y = output_y * stride_height +
+                                       f_y * dilation_height -
+                                       input_padding_top;
+          if (0 <= input_y && input_y < input_height) {
+            for (std::int32_t f_x = 0; f_x < filter_width; f_x++) {
+              const std::int32_t input_x = output_x * stride_width +
+                                           f_x * dilation_width -
+                                           input_padding_left;
+              const std::int32_t kernel_index = f_y * filter_width + f_x;
+              const std::int32_t index = output_tile_start * filter_size +
+                                         kernel_index * block_size_pixels +
+                                         output_tile_offset;
+              if (0 <= input_x && input_x < input_width) {
+                indirection_buffer.at(index) =
+                    (input_ptr + (batch_index * input_height * input_width +
+                                  input_y * input_width + input_x) *
+                                     input_depth);
+              } else {
+                indirection_buffer.at(index) = zero_buffer.data();
+              }
+            }
+          } else {
+            for (std::int32_t f_x = 0; f_x < filter_width; f_x++) {
+              const std::int32_t kernel_index = f_y * filter_width + f_x;
+              const std::int32_t index = output_tile_start * filter_size +
+                                         kernel_index * block_size_pixels +
+                                         output_tile_offset;
+              indirection_buffer.at(index) = zero_buffer.data();
+            }
+          }
+        }
+      }
+    }
   }
-}
+
+  // To be implemented by concrete subclasses.
+  virtual void Run(const std::int32_t pixel_start, const std::int32_t pixel_end,
+                   void* output_ptr) const = 0;
+
+  void Dispatch(void* output_ptr) const {
+    // TODO: implement multithreading here.
+    Run(0, num_output_pixels, output_ptr);
+  };
```

Comment:
`Run` is a virtual function that is overriden in derived classes, and performs the computation on some subset of input (range of output pixels).

`Dispatch` is what the op calls to run the kernel, which just calls `Run` for now, but in the future could use multiple threads, et cetera.